library/hdfs-library/src/test/java/uk/gov/gchq/gaffer/hdfs/integration/loader/AddElementsFromHdfsLoaderIT.java

```diff
@@ -128,18 +127,17 @@ public void _setup() throws Exception {
         }
 
         @Test
-        public void shouldThrowExceptionWhenAddElementsFromHdfsWhenFailureDirectoryContainsFiles(TestInfo testInfo) throws Exception {
+        public void shouldThrowExceptionWhenAddElementsFromHdfsWhenFailureDirectoryContainsFiles(final TestInfo testInfo) throws Exception {
             tearDown();
             fs.mkdirs(new Path(failureDir));
             try (final BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(fs.create(new Path(failureDir + "/someFile.txt"), true)))) {
                 writer.write("Some content");
             }
 
             try {
-                setup(testInfo);
-                fail("Exception expected");
-            } catch (final OperationException e) {
-                assertThat(e.getCause().getMessage()).isEqualTo("Failure directory is not empty: " + failureDir);
+                assertThatExceptionOfType(Exception.class)
+                        .isThrownBy(() -> setup(testInfo))
+                        .withStackTraceContaining("Failure directory is not empty: " + failureDir);
             } finally {
                 tearDown();
             }
```

Comment:
This can be further improved by removing the `try` and `finally` entirely. The `assertThatExceptionOfType` doesn't throw anything, meaning the `try` isn't needed, therefore the `finally` is not needed either.

Reply:
To be clear, the `tearDown()` should be kept, it's the `finally` surrounding it that can be removed.

Reply:
If the test fails will the `tearDown` still be called if it is in a `finally` block? If it fails and the `tearDown` is not in a `finally` block then I figure it will not be. Seems like it should be in an `@AfterEach` method instead...

Reply:
You're correct. If the assertion fails, then the lines below will not be called. I'm not sure if this throws an exception though, it may stop execution using another method which would prevent the `finally` from being run.

The good news is that is you dig down to the class where `tearDown` is declared, it turns out this is executed `@AfterAll` tests anyway:
https://github.com/gchq/Gaffer/blob/9986d3b0dc54e95fcaf24128758da4d23c793954/integration-test/src/test/java/uk/gov/gchq/gaffer/integration/AbstractStoreIT.java#L186-L189

The calls to `tearDown` at the end of the tests in this class are redundant and can be removed entirely as JUnit is already calling this method after every test method. I'm not sure why it is being called at the start of the method either, as it will have just been called at the end of the previous test. No other test methods in the codebase do this.

Reply:
`tearDown` is called by 2 test methods so would need to be put in a `@AfterEach` block instead/as well?

Reply:
I was getting mixed up between `@AfterAll` and `@AfterEach`. Adding an `@AfterEach` is required if we want to ensure it always runs, even if the assertion fails.

However, quite why this class needs to call `tearDown` or `setup` is unclear. Calling `tearDown` simply sets the graph to `null` and `setup` then reconfigures it. Before each test is run `setup` is called anyway, so why set it to `null` and then call `setup` if `setup` was already called before the method was run?

For now I think doing as you have suggested is the best approach. Improving this class to remove the apparently redundant `tearDown` or `setup` should be a separate issue.